From patchwork Fri May 17 08:49:47 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10947523
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 769451398
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:17 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6553E228C8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:17 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 597DE22A2A; Fri, 17 May 2019 08:50:17 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 05416228C8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:17 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728023AbfEQIt4 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 17 May 2019 04:49:56 -0400
Received: from mail-pg1-f196.google.com ([209.85.215.196]:43617 "EHLO
        mail-pg1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727624AbfEQIt4 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 17 May 2019 04:49:56 -0400
Received: by mail-pg1-f196.google.com with SMTP id t22so2977662pgi.10;
        Fri, 17 May 2019 01:49:56 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:mime-version
         :content-transfer-encoding;
        bh=9I+lX7qkOYzggWtQec+1m6I5euArD7SEmoWKuSnERRI=;
        b=PaW707s8aG5UEv/PL/RpVgZMRPo/yXTGiv/I0BSmqeqLzCCzlveC2OIIVi79sHPSaH
         tM6P+W5oVQbCKsTjX7CL1gOk0fCIzCjO0CacoL6mzR9QT7ueQj4pg5Ump9BZkwbbl+1t
         a2D8p/tXSYhyJZOtIlID2Tcwng1SolxVJgCEdwq0nBCbJw7DNcuS/UMGL8C0TUb6BFPl
         LjRI3LcYt84a5Uq48zn4KXpI2K6mVMi+4OmRwzEs4nslVmHKWJF8rshg+U3YcMDXs7ZF
         FQW2zgzMck7acr9+tPQnwsuDEUkYYT/lQdJ8NxiFG6Scpx006rXv3N9CtaHICpVIf7EV
         /arg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:mime-version
         :content-transfer-encoding;
        bh=9I+lX7qkOYzggWtQec+1m6I5euArD7SEmoWKuSnERRI=;
        b=oyFooB4UPaqNbv2DyjQlIDAmTwxkD6D2ALjSuK06s4QBSCPASCaKz0/yDBFQj8IWC0
         hrJ23izyFH5QB6kGpRXrH+5x8VXxkuCbjJX0whKUv6Ltm7ZH00iaXnnYlgCWvyIWQE22
         wfpWoG3Q4VWMjHgCPESvXJP7DqYmmU05pSMg+v6TC0/YOVwbvAfaeRysClHfXvtTxtJb
         Nv7h2N0mcK/oPLAgaDTnbFvkdWRK5LRUfJQBZRm8uuIVHcwc16y/Xfvx3TvxIqt/UmjZ
         RyaFj36j0EMQycM8qa0SeqoF1VPQGj8DQxVxDP0PMse0V2OVc/Xy8YyvV4WA/O+r1BhH
         pSNg==
X-Gm-Message-State: APjAAAVz7phDLUMsVckfa8g/QCoo3Oyx45dOykr9hH8EhvT0dP3Rvqxg
        R3QmqXOmf4Y30rDj1sCfw+5ZeIgI
X-Google-Smtp-Source: 
 APXvYqwFYB1b7T68JzkdwK3+oTP4wh54dJfjJWuzNZcQtn+5UzG7e3oluVqdg4LO8ooo1ku5XcM3Ug==
X-Received: by 2002:a63:e408:: with SMTP id a8mr55686792pgi.146.1558082995724;
        Fri, 17 May 2019 01:49:55 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 63sm10417127pfu.95.2019.05.17.01.49.53
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Fri, 17 May 2019 01:49:54 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Liran Alon <liran.alon@oracle.com>
Subject: [PATCH 1/4] KVM: x86: Disable intercept for CORE cstate read
Date: Fri, 17 May 2019 16:49:47 +0800
Message-Id: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Allow guest reads CORE cstate when exposing host CPU power management capabilities 
to the guest. PKG cstate is restricted to avoid a guest to get the whole package 
information in multi-tenant scenario.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Sean Christopherson <sean.j.christopherson@intel.com>
Cc: Liran Alon <liran.alon@oracle.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kvm/vmx/vmx.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index 771d3bf..b0d6be5 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -6615,6 +6615,12 @@ static struct kvm_vcpu *vmx_create_vcpu(struct kvm *kvm, unsigned int id)
 	vmx_disable_intercept_for_msr(msr_bitmap, MSR_IA32_SYSENTER_CS, MSR_TYPE_RW);
 	vmx_disable_intercept_for_msr(msr_bitmap, MSR_IA32_SYSENTER_ESP, MSR_TYPE_RW);
 	vmx_disable_intercept_for_msr(msr_bitmap, MSR_IA32_SYSENTER_EIP, MSR_TYPE_RW);
+	if (kvm_mwait_in_guest(kvm)) {
+		vmx_disable_intercept_for_msr(msr_bitmap, MSR_CORE_C1_RES, MSR_TYPE_R);
+		vmx_disable_intercept_for_msr(msr_bitmap, MSR_CORE_C3_RESIDENCY, MSR_TYPE_R);
+		vmx_disable_intercept_for_msr(msr_bitmap, MSR_CORE_C6_RESIDENCY, MSR_TYPE_R);
+		vmx_disable_intercept_for_msr(msr_bitmap, MSR_CORE_C7_RESIDENCY, MSR_TYPE_R);
+	}
 	vmx->msr_bitmap_mode = 0;
 
 	vmx->loaded_vmcs = &vmx->vmcs01;

From patchwork Fri May 17 08:49:48 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10947521
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 09D981398
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:15 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id ED5DF228C8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:14 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id E17C322A2A; Fri, 17 May 2019 08:50:14 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 7902F228C8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:14 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728291AbfEQIt7 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 17 May 2019 04:49:59 -0400
Received: from mail-pg1-f194.google.com ([209.85.215.194]:33230 "EHLO
        mail-pg1-f194.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727624AbfEQIt6 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 17 May 2019 04:49:58 -0400
Received: by mail-pg1-f194.google.com with SMTP id h17so3005949pgv.0;
        Fri, 17 May 2019 01:49:58 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=BMZTqCRe7Yq2IM+sICYMlznnn3t4JBojU8sT9FmlO4o=;
        b=cW7xcpKhrqtkIzyTFrMiDQ5p+TtZrCS3gBiD2zsnKHX/876EjpglYdiYy2K5mBdSzw
         b1J3wBqorDacpSoiBdeGkqVb+vsjpcRbAeW0ctnvw+yrzavMsJ2R/2dF6vvvoNrYPYX8
         CO5uZWmCMd67CPIxJbIfTM+FcpP84IKzIMjpYW7eBcI/MTmETqqOZf7Er8SltTqWLepa
         RbfWJ4+MHMCH3MmujNL/PFp2BKsuG323gwwPja4JmB908c8fVEJT5eUHR96WaEZVzZOI
         hOnjRqjgs3lqBQiHPBUDQuRopPoYCnfeQD2OKioIgzBRgv82jd0Cl2qJS7br2WoX5Ktp
         Z9jA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=BMZTqCRe7Yq2IM+sICYMlznnn3t4JBojU8sT9FmlO4o=;
        b=FUHYNYA44qyb8i2ryx8qAaDKZf5wZ0f9wr8MyjQZoSnimi2CZPhzJ10atY2+n6xiui
         X5kyKfPSdSOMt4db1u8fpRErsLbyi08OzLZPA+sJqFhOJzMaUiNifD4bG4DIntfdvYhW
         gWO9VIPUzhZAl1pMhaFIhIHQK/0n/Y5H9BepwBRW0y2nmmB5/izAAWNjOAV6mWLLI29O
         0VxL1Z4O6BUg7xsswJqtJTjhn10cE9krgvfZCXRtiYeYASOQk8TWVzIT6kTQHSqPdT5z
         ELsolKnnLLMAe2XRGj5yWS7KtTIOwd/+P9KyI5T8/2w6xkdlBoKN8puDWHCpKbPCTFWO
         vKog==
X-Gm-Message-State: APjAAAWFwJ62bLR57jb0DutRALS3dmzGFrV954weEFkGctm7JvJ3j5X9
        4HyUOUdduKgLo459UZIlhEUfRICG
X-Google-Smtp-Source: 
 APXvYqw39p1zjnYKnmyqGgifghpRVUZmXaCSxgucKvVBhasmkJ0xJcGLVC/TXH1ct3+/qFkrArIAsg==
X-Received: by 2002:aa7:9a99:: with SMTP id
 w25mr20416545pfi.249.1558082998095;
        Fri, 17 May 2019 01:49:58 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 63sm10417127pfu.95.2019.05.17.01.49.55
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Fri, 17 May 2019 01:49:57 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Liran Alon <liran.alon@oracle.com>,
 Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Subject: [PATCH RESEND 2/4] KVM: X86: Emulate MSR_IA32_MISC_ENABLE MWAIT bit
Date: Fri, 17 May 2019 16:49:48 +0800
Message-Id: <1558082990-7822-2-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
References: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

MSR IA32_MSIC_ENABLE bit 18, according to SDM:

| When this bit is set to 0, the MONITOR feature flag is not set (CPUID.01H:ECX[bit 3] = 0).
| This indicates that MONITOR/MWAIT are not supported.
|
| Software attempts to execute MONITOR/MWAIT will cause #UD when this bit is 0.
|
| When this bit is set to 1 (default), MONITOR/MWAIT are supported (CPUID.01H:ECX[bit 3] = 1).

The CPUID.01H:ECX[bit 3] ought to mirror the value of the MSR bit,
CPUID.01H:ECX[bit 3] is a better guard than kvm_mwait_in_guest().
kvm_mwait_in_guest() affects the behavior of MONITOR/MWAIT, not its
guest visibility.

This patch implements toggling of the CPUID bit based on guest writes
to the MSR.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Sean Christopherson <sean.j.christopherson@intel.com>
Cc: Liran Alon <liran.alon@oracle.com>
Cc: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kvm/cpuid.c | 8 ++++++++
 arch/x86/kvm/x86.c   | 9 +++++++++
 2 files changed, 17 insertions(+)

diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index fd39516..9244d63 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -137,6 +137,14 @@ int kvm_update_cpuid(struct kvm_vcpu *vcpu)
 		(best->eax & (1 << KVM_FEATURE_PV_UNHALT)))
 		best->eax &= ~(1 << KVM_FEATURE_PV_UNHALT);
 
+	best = kvm_find_cpuid_entry(vcpu, 0x1, 0);
+	if (best) {
+		if (vcpu->arch.ia32_misc_enable_msr & MSR_IA32_MISC_ENABLE_MWAIT)
+			best->ecx |= F(MWAIT);
+		else
+			best->ecx &= ~F(MWAIT);
+	}
+
 	/* Update physical-address width */
 	vcpu->arch.maxphyaddr = cpuid_query_maxphyaddr(vcpu);
 	kvm_mmu_reset_context(vcpu);
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 1d89cb9..f2e3847 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -2506,6 +2506,15 @@ int kvm_set_msr_common(struct kvm_vcpu *vcpu, struct msr_data *msr_info)
 		}
 		break;
 	case MSR_IA32_MISC_ENABLE:
+		if ((vcpu->arch.ia32_misc_enable_msr ^ data) & MSR_IA32_MISC_ENABLE_MWAIT) {
+			if ((vcpu->arch.ia32_misc_enable_msr & MSR_IA32_MISC_ENABLE_MWAIT) &&
+				!(data & MSR_IA32_MISC_ENABLE_MWAIT)) {
+				if (!guest_cpuid_has(vcpu, X86_FEATURE_XMM3))
+					return 1;
+			}
+			vcpu->arch.ia32_misc_enable_msr = data;
+			kvm_update_cpuid(vcpu);
+		}
 		vcpu->arch.ia32_misc_enable_msr = data;
 		break;
 	case MSR_IA32_SMBASE:

From patchwork Fri May 17 08:49:49 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10947519
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id C001E924
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:11 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AE50D228C9
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:11 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id A23C422A65; Fri, 17 May 2019 08:50:11 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 4A260228C9
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:11 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728508AbfEQIuB (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 17 May 2019 04:50:01 -0400
Received: from mail-pl1-f193.google.com ([209.85.214.193]:35521 "EHLO
        mail-pl1-f193.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728481AbfEQIuA (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 17 May 2019 04:50:00 -0400
Received: by mail-pl1-f193.google.com with SMTP id g5so3056558plt.2;
        Fri, 17 May 2019 01:50:00 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=5VREHkPw8R8lwRxH+KVXuUktO+8MCxC49x2eL5UCIow=;
        b=RPF5qcECkRQX0J004+yKvh9PHjuDP9QRAl1CTIGwYdfFjC73ZOLGRcxDyItiIW7XyA
         nbv2F1mtgRdvb5ihxRpkb3z/RtFdWaS3jZoQtTHtuZuaN+M96WdPShDILNxokcxMGmK7
         fRmgTv9Wt0N7ybh4BTAndZqH7yw/uUDNoe0JHoE3xA7owQGGbmrO/zzIsC10J4WDrEyJ
         WSfQKHZF4soeugQiLr5Qp2Kk1HO4TyKa5EWRgd2eL1cm66YKg1l6GYzikIeHsFV6SVXG
         cMYlPysB335VCYVBaSNNenhR8jQkxOVjEHMf0XmlklLT+MhhJ7V4gRkvDoXoX9NGDoMm
         60dA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=5VREHkPw8R8lwRxH+KVXuUktO+8MCxC49x2eL5UCIow=;
        b=TI1hwb2e8AuC8zzKWCHsFfPu3lqRQKdhO8xu1tWvk3LyOFbF4s+n5riZCwkm2s75Pc
         mKdKy3y2eNXFutAVycF5Wx9aTMfQ80uKlu3g8ZYtTz7J9AflsHHAwwmShbgIFuMVbX5Y
         1ZRAUewCLVAcJJjtfgfAkuGxIGOuZboQLfKqe7UN42DivCQXs/fO+O7CA65iHPHWo9QQ
         ZY2Dk3jWdwRGlgTxc+EApwJHEWrpoYVZRQQgr1LU4XNHJlNW5KpSOUl6Pt4a/7+UDv/E
         6ds7oe2QyBpiX6/tRxJFvMiOtTXV7uZNAz/ThrOSE1xlPptkUoSMW2GjgMPBQs5K8liq
         jccA==
X-Gm-Message-State: APjAAAXjW/PQL+Kt8nqjjXn7Dh32PRx6yjdiYxGCc3fIpxYUkDgKqva3
        gHhBEU36gnQx1lkiGYAMTOAfrOIe
X-Google-Smtp-Source: 
 APXvYqxe4n7YAE9tLJpT4f6BzKilEwXE3tVIRNBySLRokH+e3XoGXrD1560o4yN2vgR74SJe88LRlA==
X-Received: by 2002:a17:902:aa95:: with SMTP id
 d21mr1072036plr.32.1558083000075;
        Fri, 17 May 2019 01:50:00 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 63sm10417127pfu.95.2019.05.17.01.49.58
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Fri, 17 May 2019 01:49:59 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>,
 "Paul E . McKenney" <paulmck@linux.ibm.com>
Subject: [PATCH 3/4] KVM: Fix spinlock taken warning during host resume
Date: Fri, 17 May 2019 16:49:49 +0800
Message-Id: <1558082990-7822-3-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
References: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

 WARNING: CPU: 0 PID: 13554 at kvm/arch/x86/kvm//../../../virt/kvm/kvm_main.c:4183 kvm_resume+0x3c/0x40 [kvm]
  CPU: 0 PID: 13554 Comm: step_after_susp Tainted: G           OE     5.1.0-rc4+ #1
  RIP: 0010:kvm_resume+0x3c/0x40 [kvm]
  Call Trace:
   syscore_resume+0x63/0x2d0
   suspend_devices_and_enter+0x9d1/0xa40
   pm_suspend+0x33a/0x3b0
   state_store+0x82/0xf0
   kobj_attr_store+0x12/0x20
   sysfs_kf_write+0x4b/0x60
   kernfs_fop_write+0x120/0x1a0
   __vfs_write+0x1b/0x40
   vfs_write+0xcd/0x1d0
   ksys_write+0x5f/0xe0
   __x64_sys_write+0x1a/0x20
   do_syscall_64+0x6f/0x6c0
   entry_SYSCALL_64_after_hwframe+0x49/0xbe

Commit ca84d1a24 (KVM: x86: Add clock sync request to hardware enable) mentioned 
that "we always hold kvm_lock when hardware_enable is called.  The one place that 
doesn't need to worry about it is resume, as resuming a frozen CPU, the spinlock 
won't be taken." However, commit 6706dae9 (virt/kvm: Replace spin_is_locked() with 
lockdep) introduces a bug, it asserts when the lock is not held which is contrary 
to the original goal. 

This patch fixes it by WARN_ON when the lock is held.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Cc: Paul E. McKenney <paulmck@linux.ibm.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 virt/kvm/kvm_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 5fb0f16..c7eab5f 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -4096,7 +4096,7 @@ static int kvm_suspend(void)
 static void kvm_resume(void)
 {
 	if (kvm_usage_count) {
-		lockdep_assert_held(&kvm_count_lock);
+		WARN_ON(lockdep_is_held(&kvm_count_lock));
 		hardware_enable_nolock(NULL);
 	}
 }

From patchwork Fri May 17 08:49:50 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10947517
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 1BE46924
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:09 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 07AD4228C8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:09 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id EA65F22A2A; Fri, 17 May 2019 08:50:08 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8A901228C8
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri, 17 May 2019 08:50:08 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728638AbfEQIuE (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 17 May 2019 04:50:04 -0400
Received: from mail-pl1-f193.google.com ([209.85.214.193]:35523 "EHLO
        mail-pl1-f193.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728566AbfEQIuC (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 17 May 2019 04:50:02 -0400
Received: by mail-pl1-f193.google.com with SMTP id g5so3056594plt.2;
        Fri, 17 May 2019 01:50:02 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=LXdojfs9tV6rnZ/fXmeiGAyGOFlNjDR2xUpvGACHON0=;
        b=EG+mPKckPfuZ2vfk8a6ttfLtJ+/laZpt7eB1TWVIm2VkpYk/ukPr+nsXXk5elLZlMD
         NV8i3ZybA4tKsIYD23t+H9z5McPcNQalHXvM/Zc359jp6nH5dT/VjUw1CiLzrKM+v62I
         zeLFcywisHAc+lMh+j3VbDjV9/LMkepD6ezDp4BJNGKjevFijsZdiYbGc9zDlwIV8qVE
         uFWxP5l27aXZgURttQRNn2O9SyabMmKmNgN4iZUuvlcM4HTklAvI5pphFEoWo//wN/0w
         hGEAoOEG+t8HaTOWTQ94DieJcjffwOkfIXVv8stpZjysMiPcpsy9S82sNZghzQuE3Ubx
         OS5A==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=LXdojfs9tV6rnZ/fXmeiGAyGOFlNjDR2xUpvGACHON0=;
        b=PnsEBpe27t6m4Q51otH5El4u5vDZbOWQD8dFbWiGLOeE3lUNlmuiSpBGqzvYQmlaNA
         c8X5HZDPLRM2vCR0PkEkrMiZf5ML3k183FKKL2JNSVK3AoiRMRY6oyqSeM6QXCrzioQi
         IRraLjgO/J805rNNOFVQWm2hlrx6s5DsjQTd8hg/vnMSklAGm2J2/EWJ8kMj4ARDzfEg
         EmhwpXkQ+H0SQlWahPO3ByiXfX/4a6NlTccDVQMsF0HxIlp4/E+puEz+m6xhzQRJPsQO
         1uYJke61LN2+f9c5iilN6WlwssifCFwWPqj+gYRWYlajtfd41CnFGIGMAIUio5rT3THW
         mRig==
X-Gm-Message-State: APjAAAUHUXtY9BlJcT8qVDNExrBOo6bBbZhdwf/fe6tvINuLn9ZwnRFT
        NRJhMyB+AO4Z1k+trDqJipnjhPws
X-Google-Smtp-Source: 
 APXvYqx6UZmeM1Meepadf6z2mpg8VSstpYtAyfG5btHRv7xogLmK8f4zFUKNMaJW9hanOFnKfnyYuQ==
X-Received: by 2002:a17:902:b492:: with SMTP id
 y18mr50545153plr.96.1558083001916;
        Fri, 17 May 2019 01:50:01 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 63sm10417127pfu.95.2019.05.17.01.50.00
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Fri, 17 May 2019 01:50:01 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>
Subject: [PATCH 4/4] KVM: nVMX: Fix using __this_cpu_read() in preemptible
 context
Date: Fri, 17 May 2019 16:49:50 +0800
Message-Id: <1558082990-7822-4-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
References: <1558082990-7822-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

 BUG: using __this_cpu_read() in preemptible [00000000] code: qemu-system-x86/4590
  caller is nested_vmx_enter_non_root_mode+0xebd/0x1790 [kvm_intel]
  CPU: 4 PID: 4590 Comm: qemu-system-x86 Tainted: G           OE     5.1.0-rc4+ #1
  Call Trace:
   dump_stack+0x67/0x95
   __this_cpu_preempt_check+0xd2/0xe0
   nested_vmx_enter_non_root_mode+0xebd/0x1790 [kvm_intel]
   nested_vmx_run+0xda/0x2b0 [kvm_intel]
   handle_vmlaunch+0x13/0x20 [kvm_intel]
   vmx_handle_exit+0xbd/0x660 [kvm_intel]
   kvm_arch_vcpu_ioctl_run+0xa2c/0x1e50 [kvm]
   kvm_vcpu_ioctl+0x3ad/0x6d0 [kvm]
   do_vfs_ioctl+0xa5/0x6e0
   ksys_ioctl+0x6d/0x80
   __x64_sys_ioctl+0x1a/0x20
   do_syscall_64+0x6f/0x6c0
   entry_SYSCALL_64_after_hwframe+0x49/0xbe

Accessing per-cpu variable should disable preemption, this patch extends the 
preemption disable region for __this_cpu_read().

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
---
 arch/x86/kvm/vmx/nested.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/vmx/nested.c b/arch/x86/kvm/vmx/nested.c
index 0c601d0..8f6f69c 100644
--- a/arch/x86/kvm/vmx/nested.c
+++ b/arch/x86/kvm/vmx/nested.c
@@ -2792,14 +2792,13 @@ static int nested_vmx_check_vmentry_hw(struct kvm_vcpu *vcpu)
 	      : "cc", "memory"
 	);
 
-	preempt_enable();
-
 	if (vmx->msr_autoload.host.nr)
 		vmcs_write32(VM_EXIT_MSR_LOAD_COUNT, vmx->msr_autoload.host.nr);
 	if (vmx->msr_autoload.guest.nr)
 		vmcs_write32(VM_ENTRY_MSR_LOAD_COUNT, vmx->msr_autoload.guest.nr);
 
 	if (vm_fail) {
+		preempt_enable();
 		WARN_ON_ONCE(vmcs_read32(VM_INSTRUCTION_ERROR) !=
 			     VMXERR_ENTRY_INVALID_CONTROL_FIELD);
 		return 1;
@@ -2811,6 +2810,7 @@ static int nested_vmx_check_vmentry_hw(struct kvm_vcpu *vcpu)
 	local_irq_enable();
 	if (hw_breakpoint_active())
 		set_debugreg(__this_cpu_read(cpu_dr7), 7);
+	preempt_enable();
 
 	/*
 	 * A non-failing VMEntry means we somehow entered guest mode with
